File: naru_agent/orchestration/handoff.py
"""AgentHandoffLoop — follows handoff chains across agents with safety limits."""

from __future__ import annotations

import logging

from naru_agent.agent import NaruResult
from naru_agent.orchestration.orchestrator import AgentChatDelegate

logger = logging.getLogger(__name__)


class AgentHandoffLoop:
    """Orchestrator wrapper that follows handoff chains with safety limits."""

    def __init__(
        self,
        agents: dict[str, AgentChatDelegate],
        *,
        entry: str,
        max_handoffs: int = 5,
        name: str = "handoff_loop",
    ) -> None:
        if not agents:
            raise ValueError("Handoff loop must have at least one agent")
        if entry not in agents:
            raise ValueError(f"Entry agent '{entry}' not found in agents")
        self._agents = agents
        self._entry = entry
        self._max_handoffs = max_handoffs
        self.name = name

    def chat(
        self,
        message: str,
        user_id: str | None = None,
        session_id: str | None = None,
    ) -> NaruResult:
        current_agent_name = self._entry
        current_message = message
        visited: list[str] = []
        result: NaruResult | None = None

        for _ in range(self._max_handoffs + 1):
            agent = self._agents.get(current_agent_name)
            if agent is None:
                raise ValueError(f"Unknown agent: {current_agent_name}")

            visited.append(current_agent_name)
            result = agent.chat(current_message, user_id=user_id, session_id=session_id)

            if result.handoff is None:
                return result

            # Follow handoff
            current_agent_name = result.handoff.target
            current_message = result.handoff.message or message

        # Limit reached — return last result
        logger.warning(
            "Handoff limit reached (%d), returning last result. Chain: %s",
            self._max_handoffs,
            " → ".join(visited),
        )
        return result  # type: ignore[return-value]

File: js/dist-cjs/orchestration/handoff.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentHandoffLoop = void 0;
/**
 * AgentHandoffLoop — follows handoff chains across agents with safety limits.
 */
class AgentHandoffLoop {
    agents;
    entry;
    maxHandoffs;
    name;
    constructor(agents, entry, maxHandoffs = 5, name = "handoff_loop") {
        this.agents = agents;
        this.entry = entry;
        this.maxHandoffs = maxHandoffs;
        this.name = name;
        if (agents.size === 0) {
            throw new Error("Handoff loop must have at least one agent");
        }
        if (!agents.has(entry)) {
            throw new Error(`Entry agent '${entry}' not found in agents`);
        }
    }
    async chat(message, options) {
        let currentAgent = this.entry;
        let currentMessage = message;
        let result = null;
        for (let i = 0; i <= this.maxHandoffs; i++) {
            const agent = this.agents.get(currentAgent);
            if (!agent) {
                throw new Error(`Unknown agent: ${currentAgent}`);
            }
            result = await agent.chat(currentMessage, options);
            if (!result.handoff) {
                return result;
            }
            currentAgent = result.handoff.target;
            currentMessage = result.handoff.message ?? message;
        }
        // Limit reached — return last result
        return result;
    }
}
exports.AgentHandoffLoop = AgentHandoffLoop;
